File: world-info-service/src/main/java/com/example/worldinfo/worldinfoservice/models/responses/pagination/Pagination.java
package com.example.worldinfo.worldinfoservice.models.responses.pagination;

import java.io.Serializable;
import java.util.List;

public abstract class Pagination implements Serializable {
    private static final long serialVersionUID = 1L;
    private final int pageNumber;
    private final int limit;
    private final int total;
    private final int offset;
    private final String orderBy;
    private final String order;
    private final List<PaginationLink> links;

    public Pagination(int pageNumber, int limit, int total, int offset, String orderBy, String order, List<PaginationLink> links) {
        this.pageNumber = pageNumber;
        this.limit = limit;
        this.total = total;
        this.offset = offset;
        this.orderBy = orderBy;
        this.order = order;
        this.links = links;
    }

    @Override
    public String toString() {
        return "Pagination{" +
                "pageNumber=" + pageNumber +
                ", limit=" + limit +
                ", total=" + total +
                ", offset=" + offset +
                ", orderBy='" + orderBy + '\'' +
                ", orderAsc=" + order +
                ", links=" + links +
                '}';
    }

    public String getOrder() {
        return order;
    }

    public String getOrderBy() {
        return orderBy;
    }

    public List<PaginationLink> getLinks() {
        return links;
    }

    public int getOffset() {
        return offset;
    }

    public int getPageNumber() {
        return pageNumber;
    }

    public int getLimit() {
        return limit;
    }

    public int getTotal() {
        return total;
    }

}

File: world-info-client/src/app/models/pagination.ts
import { HttpParams } from '@angular/common/http';
import {
  ICountry,
  ICountryPagination,
  IPagination,
  IPaginationLink,
} from '../interfaces/interfaces';

export abstract class Pagination implements IPagination {
  pageNumber: number;
  limit: number;
  total: number;
  offset: number;
  orderBy: string;
  order: string;
  links: IPaginationLink[];

  constructor() {
    this.pageNumber = 0;
    this.limit = 10;
    this.total = 0;
    this.offset = 0;
    this.orderBy = 'name';
    this.order = 'DESC';
    this.links = [];
  }

  setPagination(pagination: ICountryPagination): void {
    this.pageNumber = pagination.pageNumber;
    this.limit = pagination.limit;
    this.total = pagination.total;
    this.offset = pagination.offset;
    this.orderBy = pagination.orderBy;
    this.order = pagination.order;
    this.links = pagination.links;
  }

  setLimit(limit: number): void {
    this.limit = limit;
  }

  setOffset(offset: number): void {
    this.offset = offset;
  }

  setOrderBy(orderBy: string): void {
    this.orderBy = orderBy;
  }

  setOrder(order: string): void {
    this.order = order;
  }

  setLinks(links: IPaginationLink[]): void {
    this.links = links;
  }

  setTotal(total: number): void {
    this.total = total;
  }

  setPageNumber(pageNumber: number): void {
    this.pageNumber = pageNumber;
  }

  getLimit(): number {
    return this.limit;
  }

  getOffset(): number {
    return this.offset;
  }

  getOrderBy(): string {
    return this.orderBy;
  }

  getOrder(): string {
    return this.order;
  }

  getLinks(): IPaginationLink[] {
    return this.links;
  }

  getTotal(): number {
    return this.total;
  }

  getPageNumber(): number {
    return this.pageNumber;
  }

  getPreviousLink(): string {
    for (const link of this.links) {
      if (link.name === 'prev') {
        return link.uri;
      }
    }
    return '';
  }

  getNextLink(): string {
    for (const link of this.links) {
      if (link.name === 'next') {
        return link.uri;
      }
    }
    return '';
  }

  getLastLink(): string {
    for (const link of this.links) {
      if (link.name === 'last') {
        return link.uri;
      }
    }
    return '';
  }

  getFirstLink(): string {
    for (const link of this.links) {
      if (link.name === 'first') {
        return link.uri;
      }
    }
    return '';
  }

  getHttpParams(): HttpParams {
    return new HttpParams()
      .set('limit', this.limit)
      .set('offset', this.offset)
      .set('property', this.orderBy)
      .set('order', this.order);
  }
}
